src/cls/sem_set/module.cc

```diff
@@ -0,0 +1,235 @@
+              __PRETTY_FUNCTION__, __LINE__, e.what());
+      return -EIO;
+    }
+
+  return 0;
+}
+
+int decrement(cls_method_context_t hctx, buffer::list *in, buffer::list *out)
+{
+  CLS_LOG(10, "%s", __PRETTY_FUNCTION__);
+
+  ss::decrement op;
+  try {
+    auto iter = in->cbegin();
+    decode(op, iter);
+  } catch (const std::exception& e) {
+    CLS_ERR("ERROR: %s:%d failed to decode request: %s", __PRETTY_FUNCTION__,
+            __LINE__, e.what());
+    return -EINVAL;
+  }
+
+  const auto& max_omap = cls_get_config(hctx).get_val<std::uint64_t>(
+    "osd_max_omap_entries_per_request"sv);
+  if (op.keys.size() > max_omap) {
+    CLS_ERR("ERROR: %s:%d too many keys: %zu", __PRETTY_FUNCTION__,
+	    __LINE__, op.keys.size());
+    return -E2BIG;
+  }
+
+  for (const auto& key_ : op.keys) try {
+      buffer::list valbl;
+      auto key = std::string(PREFIX) + key_;
+      auto r = cls_cxx_map_get_val(hctx, key, &valbl);
+      if (r < 0) {
+        CLS_ERR("ERROR: %s:%d failed to read semaphore: r=%d",
+                __PRETTY_FUNCTION__, __LINE__, r);
+        return r;
+      }
+      auto sem = from_bl<std::uint64_t>(valbl);
+      if (sem > 1) {
+	--sem;
+        valbl.clear();
+        ceph::encode(sem, valbl);
+        r = cls_cxx_map_set_val(hctx, key, &valbl);
+        if (r < 0) {
+          CLS_ERR("ERROR: %s:%d failed to update semaphore: r=%d",
+                  __PRETTY_FUNCTION__, __LINE__, r);
+          return r;
+        }
+      } else {
+        r = cls_cxx_map_remove_key(hctx, key);
+        if (r < 0) {
+          CLS_ERR("ERROR: %s:%d failed to remove key %s: r=%d",
+                  __PRETTY_FUNCTION__, __LINE__, key.c_str(), r);
+          return r;
+        }
+      }
+    } catch (const std::exception& e) {
+      CLS_ERR("CORRUPTION: %s:%d failed to decode semaphore: %s",
+              __PRETTY_FUNCTION__, __LINE__, e.what());
+      return -EIO;
+    }
+
+  return 0;
+}
+
+int list(cls_method_context_t hctx, buffer::list *in, buffer::list *out)
+{
+  CLS_LOG(10, "%s", __PRETTY_FUNCTION__);
+
+  ss::list_op op;
+  try {
+    auto iter = in->cbegin();
+    decode(op, iter);
+  } catch (const std::exception& e) {
+    CLS_ERR("ERROR: %s:%d failed to decode request: %s", __PRETTY_FUNCTION__,
+            __LINE__, e.what());
+    return -EINVAL;
+  }
+
+  auto count = op.count;
+  const auto& max_omap = cls_get_config(hctx).get_val<std::uint64_t>(
+    "osd_max_omap_entries_per_request"sv);
+  if (count > max_omap) {
+    // We clamp rather than error, here, since a lister can just
+    // continue listing with the supplied cursor.
+    count = max_omap;
+  }
+  ss::list_ret res;
+  res.cursor = op.cursor;
+
+  while (count > 0) {
+    std::map<std::string, ceph::buffer::list> vals;
+    bool more = false;
+    auto r = cls_cxx_map_get_vals(hctx,
+				  res.cursor ? *res.cursor : std::string{},
+				  std::string(PREFIX), count,
+				  &vals, &more);
```

Comment:
This is a leftover from when I was doing out-of-line deletion. I want to do inline first, then after we have everything debugged and know it's working, try switching to out-of-line to see if it improves performance. If we're interested in doing out-of-line I'd like to keep this so it can be slotted in easily, though if you think it's not worth doing I'll ditch it.